toolchain/check/member_access.cpp

```diff
@@ -458,49 +457,61 @@ static auto ValidateTupleIndex(Context& context, SemIR::LocId loc_id,
+    return HandleAction<SemIR::AccessMemberAction>(
+        context, loc_id,
+        {.type_id = SemIR::InstType::SingletonTypeId,
+         .base_id = base_id,
+         .name_id = name_id});
+  } else {
+    return HandleAction<SemIR::AccessOptionalMemberAction>(
+        context, loc_id,
+        {.type_id = SemIR::InstType::SingletonTypeId,
+         .base_id = base_id,
+         .name_id = name_id});
+  }
 }
 
-auto PerformAction(Context& context, SemIR::LocId loc_id,
-                   SemIR::AccessMemberAction action) -> SemIR::InstId {
-  SemIR::InstId base_id = action.base_id;
-  SemIR::NameId name_id = action.name_id;
-
+// Common logic for `AccessMemberAction` and `AccessOptionalMemberAction`.
+static auto PerformActionHelper(Context& context, SemIR::LocId loc_id,
+                                SemIR::InstId base_id, SemIR::NameId name_id,
+                                bool required) -> SemIR::InstId {
   // If the base is a name scope, such as a class or namespace, perform lookup
   // into that scope.
   if (auto base_const_id = context.constant_values().Get(base_id);
       base_const_id.is_constant()) {
     llvm::SmallVector<LookupScope> lookup_scopes;
     if (AppendLookupScopesForConstant(context, loc_id, base_const_id,
                                       &lookup_scopes)) {
-      return LookupMemberNameInScope(context, loc_id, base_id, name_id,
-                                     base_const_id, lookup_scopes,
-                                     /*lookup_in_type_of_base=*/false);
+      return LookupMemberNameInScope(
+          context, loc_id, base_id, name_id, base_const_id, lookup_scopes,
+          /*lookup_in_type_of_base=*/false, /*required=*/required);
     }
   }
 
   // If the base isn't a scope, it must have a complete type.
   auto base_type_id = context.insts().Get(base_id).type_id();
-  if (!RequireCompleteType(
-          context, base_type_id, context.insts().GetLocId(base_id), [&] {
-            CARBON_DIAGNOSTIC(
-                IncompleteTypeInMemberAccess, Error,
-                "member access into object of incomplete type {0}",
-                TypeOfInstId);
-            return context.emitter().Build(
-                base_id, IncompleteTypeInMemberAccess, base_id);
-          })) {
-    return SemIR::ErrorInst::SingletonInstId;
+  auto base_loc_id = context.insts().GetLocId(base_id);
+  if (required) {
+    if (!RequireCompleteType(context, base_type_id, base_loc_id, [&] {
+          CARBON_DIAGNOSTIC(IncompleteTypeInMemberAccess, Error,
+                            "member access into object of incomplete type {0}",
+                            TypeOfInstId);
+          return context.emitter().Build(base_id, IncompleteTypeInMemberAccess,
+                                         base_id);
+        })) {
+      return SemIR::ErrorInst::SingletonInstId;
+    }
+  } else {
+    if (!TryToCompleteType(context, base_type_id, base_loc_id)) {
+      return SemIR::ErrorInst::SingletonInstId;
```

Comment:
We shouldn't refer to an `ErrorInst` if we've not produced an error message; code elsewhere will interpret this as saying an error has been produced and bail out.

Perhaps we should have a new singleton instruction ID to represent that we succeeded but found nothing. Or maybe we can add a singleton instance ID for a trivial destructor, and make name lookup for a destructor name find that as a fallback if it doesn't find anything else (removing the need for the "required" flag).

Reply:
I think I was thinking of this too much like a `diagnose` flag. I've stripped it down to return `None` for the "not found" case. Does that work better for you?

> Or maybe we can add a singleton instance ID for a trivial destructor

I'd considered this, but it felt like more of a divergence in the common case of _not_ looking for the trivial destructor. Also, I think we might want something similar for implicit copy/move support.